common/src/main/java/org/mcaccess/minecraftaccess/features/EffectNarration.java

```diff
@@ -0,0 +1,86 @@
+package org.mcaccess.minecraftaccess.features;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.mcaccess.minecraftaccess.MainClass;
+import org.mcaccess.minecraftaccess.utils.NarrationUtils;
+import org.mcaccess.minecraftaccess.utils.WorldUtils;
+
+import lombok.Getter;
+import net.minecraft.client.resource.language.I18n;
+import net.minecraft.entity.effect.StatusEffectInstance;
+
+public class EffectNarration {
+    @Getter
+    private static EffectNarration instance = new EffectNarration();
+
+    private List<StatusEffectInstance> previousEffects = new ArrayList<>();
+
+    public void update() {
```

Comment:
This whole thing can be simplified by using Sets. gained = new - old, lost = old - new.

Even better would be to add an event handler (if such an event exists) or inject into where effects are added or removed.